src/time/Waveform.hpp

```diff
@@ -0,0 +1,99 @@
+#pragma once
+
+#include <Eigen/Core>
+#include "logging/Logger.hpp"
+
+namespace precice {
+
+namespace testing {
+// Forward declaration to friend the boost test struct
+class WaveformFixture;
+} // namespace testing
+
+namespace time {
+
+class Waveform {
+  friend class testing::WaveformFixture; // Make the fixture friend of this class
+public:
+  /**
+   * @brief Waveform object which stores values of current and past time windows for performing interpolation. 
+   *
+   * Storage still needs to be initialized with Waveform::initialize, before the Waveform can be used.
+   *
+   * @param interpolationOrder defines the maximum interpolation order supported by this Waveform and reserves storage correspondingly
+   */
+  Waveform(const int interpolationOrder);
+
+  /**
+   * @brief Used to initialize _timeWindowsStorage according to required size.
+   * @param valuesSize defines how many values one sample in time consists of
+   */
+  void initialize(const int valuesSize);
+
+  /**
+   * @brief Updates entry in _timeWindows corresponding to this window with given values
+   * @param values new sample for this time window
+   */
+  void store(const Eigen::VectorXd &values);
+
+  /**
+   * @brief Updates entry in _timeWindowsStorage corresponding to a given column ID with given values
+   * @param values new sample for this time window
+   * @param columnID ID of column to be updated
+   */
+  void storeAt(const Eigen::VectorXd values, int columnID);
+
+  /**
+   * @brief Called, when moving to the next time window. All entries in _timeWindows are shifted. The new entry is initialized as the value from the last window (= constant extrapolation)
+   */
+  void moveToNextWindow();
+
+  /**
+   * @brief sample Waveform. Uses interpolation with Waveform's interpolation order, if necessary
+   * @param normalizedDt time where the sampling inside the window happens. 0 refers to the beginning of the window and 1 to the end.
+   */
+  Eigen::VectorXd sample(const double normalizedDt);
+
+  // @todo try to make private!
+  /**
+   * @brief returns number of values per sample in time stored by this waveform
+   */
+  int valuesSize();
+
+  // @todo try to make private!
+  /**
+   * @brief returns number samples in time this waveform can store
+   */
+  int sizeOfSampleStorage();
+
+  // @todo try to make private!
+  /**
+   * @brief getter for a sample in _timeWindowsStorage
+   */
+  Eigen::VectorXd getSample(int sampleID);
```

Comment:
I will take care of some of the points mentioned. Encapsulation is WIP, but if I have to keep it public I will document the reason. I will extend the todos above correspondingly to not forget this.

Reply:
Fixed the todo and documentation in 9a00c8e95a3d0a26d82b4d99b843d76a77f8fcd3. I also removed the argument `sampleID`, because it is not needed.

I renamed `sizeOfSampleStorage` to `maxNumberOfStoredSamples` and replaced other occurrences of `columnID` with `sampleIndex` in f5f41f15126d419e3e2b6e5d9fd9222136566fc2.